src/array.jl

```diff
@@ -280,56 +280,118 @@ end
         index_out = offset_out + threadIdx().x
 
-        if 1 <= index_out <= length(output)
+        if index_out in range
             index_shared = threadIdx().x
-            @inbounds output[index_out] = shared[index_shared]
+            @inbounds data[index_out] = shared[index_shared]
         end
 
         return
     end
 
-    @cuda threads=nthreads blocks=nblocks shmem=shmem kernel(input, output)
+    nthreads = 256
+    nblocks = ceil(Int, length(range) / nthreads)
+    shmem = nthreads * sizeof(T)
+
+    @cuda threads=nthreads blocks=nblocks shmem=shmem kernel(data, range)
 
-    return
+    return data
 end
 
-function Base.reverse!(v::CuVector, start=1, stop=length(v))
-    v′ = view(v, start:stop)
-    _reverse(v′, v′)
-    return v
+# out-of-place
+function Base.reverse(input::CuVector{T}, start=1, stop=length(input)) where {T}
+    range = start:stop
+
+    output = similar(input)
+    start > 1 && copyto!(output, 1, input, 1, start-1)
+    stop < length(input) && copyto!(output, stop+1, input, stop+1)
+
+    function kernel(input::CuDeviceVector{T}, output::CuDeviceVector{T}, range::UnitRange) where {T}
+        index = (blockIdx().x - 1) * blockDim().x + threadIdx().x
+
+        index_in = first(range) + index - 1
+        index_out = last(range) - index + 1
+
+        if index_in <= last(range)
+            @inbounds output[index_out] = input[index_in]
+        end
+
+        return
+    end
+
+    nthreads = 256
+    nblocks = ceil(Int, length(range) / nthreads)
+    shmem = nthreads * sizeof(T)
+
+    @cuda threads=nthreads blocks=nblocks shmem=shmem kernel(input, output, range)
+
+    return output
 end
 
-function Base.reverse(v::CuVector, start=1, stop=length(v))
-    v′ = similar(v)
-    start > 1 && copyto!(v′, 1, v, 1, start-1)
-    _reverse(view(v, start:stop), view(v′, start:stop))
-    stop < length(v) && copyto!(v′, stop+1, v, stop+1)
-    return v′
+# n-dimensional
+
+function Base.reverse(input::CuArray{T, N}; dims::Integer) where {T, N}
+    if !(1 ≤ dims ≤ length(size(input)))
+      ArgumentError("dimension $dims is not 1 ≤ $dims ≤ $length(size(input))")
+    end
+
+    output = similar(input)
+
+    shape = [size(input)...]
+    numelemsinprevdims = prod(shape[1:dims-1])
+    numelemsincurrdim = shape[dims]
+
+    function kernel(input::CuDeviceArray{T, N}, output::CuDeviceArray{T, N}) where {T, N}
+        # treat an ND array as 1D
+        #
+        # after reversing by the xth dimension an element at pos
+        # [i1, i2, i3, ... , i{x},            ..., i{n}] will be at
+        # [i1, i2, i3, ... , d{x} - i{x} + 1, ..., i{n}], where
+        # d{x} is the size of the xth dimension
+
+        offset_in = blockDim().x * (blockIdx().x - 1)
+        index_in = offset_in + threadIdx().x
+
+        # the index of an element in the original array along dimension that we will flip
+        ik = ((UInt32(ceil(index_in / numelemsinprevdims)) - 1) % numelemsincurrdim) + 1
```

Comment:
@kraftpunk97 why did you use `UInt32` here? We typically do our indexing in with `Int`s.

Reply:
I thought that it was a safe move to convert indexes to UInt32, so that nothing untoward happened. The C examples that I studied used `unsigned int` indexes. Is that not the case? Sorry, I'm still learning CUDA.

Reply:
No that's fine, in (CUDA) C it's common to use such specific types, but we typically leave those off in Julia unless there's an explicit reason (performance, correctness).

Reply:
@maleadt I'm so sorry I missed this before. The `UInt32` was put there for one more reason. `ceil(index_in / numelemsinprevdims)` returns a `Float64` value, which would make `ik` and `index_out` `Float64` values. This `index_out` is being used as an index values and indices can not be Floats. This is probably the reason why the tests are failing at `@inbounds output[index_out] = input[index_in]`.

Replacing the current  `ceil` call with `ceil(Int, index_in / numelemsinprevdims...` should do the trick.

Reply:
Ah of course, I was too focused on the type not to notice the call to `ceil` :smile: